### ISSUE TITLE
diff-hl-revert-hunk doesn't close window on success

### ISSUE THREAD

**Issue by m-cat** [dgutov/diff-hl#125]
First of all, thanks for this package. It's one of the most useful Emacs packages and I love `diff-hl-amend-mode`.

My issue is as the title states. Strangely, I don't have this problem unless I set `display-buffer-base-action`. Here's my full `display-buffer` config:

```elisp
      ;; Where should we open new buffers by default?
      display-buffer-base-action '(display-buffer-below-selected)
      ;; Specify custom behavior for misbehaving buffers.
      display-buffer-alist
      '(("\\*Help\\*"
         (display-buffer-reuse-window
          display-buffer-below-selected))
        ("\\*diff-hl\\*"
         (display-buffer-same-window))
        )
```

If I answer `n`, then the new window gets properly closed, but answering `y` leaves the window open (though killing the `diff-hl` buffer, so I end up with two wndows showing the same buffer):

https://www.dropbox.com/s/kqtv1zrlfzroazg/Screen%20Shot%202019-10-09%20at%2011.07.28.png?dl=0

I've been able to determine that this happens because of something in `diff-apply-hunk` -- if I comment out this function call then the new window gets properly deleted. Haven't had time to investigate this further.

Edit: forgot to mention that I haven't had this issue with any other packages or modes. I think `diff-apply-hunk` does something weird to the window configuration, preventing it from being properly cleaned up by `quit-windows-on`.

**Comment by dgutov** [comment]
This will require some investigation. The `display-buffer` inner workings are pretty complex, and I'm not a pro on that subject.

If you find an easy fix, though, please let me know. You can ask at one of Emacs's mailing lists, too.

> I think diff-apply-hunk does something weird to the window configuration

The only thing I can see is that is calls `save-window-excursion`, which is fairly normal.

**Comment by m-cat** [comment]
Looks like I found the cause. Commenting out this line in `diff-apply-hunk` in `vc/diff-mode.el` fixes my issue:

```elisp
      ;; Display BUF in a window
      (set-window-point (display-buffer buf) (+ (car pos) (cdr new)))
```

It seems the problem is not with `diff-hl`, but with this function in `diff-mode.el` which opens a new window containing the source buffer. (Thought this would take a while to investigate, but I found this almost immediately - the comment was a big clue).

The full fix in `diff-hl` is to add a `save-window-excursion` here:

```elisp
                (let ((diff-advance-after-apply-hunk nil))
                  (save-window-excursion
                    (diff-apply-hunk t)))
```

I can't think of any downsides to doing so, and it seems to have solved my issue 100%. I have found no bad effects, having tested various `display-buffer` settings as well.
